stringdtype/stringdtype/src/dtype.c

```diff
@@ -20,19 +20,30 @@ new_stringdtype_instance(PyObject *na_object, int coerce)
 
     Py_XINCREF(na_object);
     ((StringDTypeObject *)new)->na_object = na_object;
+    npy_static_string na_name = NPY_EMPTY_STRING;
+    npy_static_string default_string = NPY_EMPTY_STRING;
     int hasnull = na_object != NULL;
     int has_nan_na = 0;
     int has_string_na = 0;
-    ss default_string = EMPTY_STRING;
     if (hasnull) {
         // first check for a string
         if (PyUnicode_Check(na_object)) {
             has_string_na = 1;
             Py_ssize_t size = 0;
             const char *buf = PyUnicode_AsUTF8AndSize(na_object, &size);
-            default_string.len = size;
-            // discards const, how to avoid?
-            default_string.buf = (char *)buf;
+            default_string = NPY_EMPTY_STRING;
+            int res = npy_string_newsize(buf, (size_t)size, &default_string);
+            if (res == -1) {
+                PyErr_NoMemory();
+                Py_DECREF(new);
+                return NULL;
+            }
+            else if (res == -2) {
+                // this should never happen
+                assert(0);
```

Comment:
It seems like this can happen if users provide nonsense for the `na_object`?!

Reply:
Maybe a helper would make sense, if we have the two different return values?  Or pass `int nogil` and set the error already in `npy_string_newsize`?
(since more than MemoryError is possible here, unless we check the size for validity first.)

Reply:
> It seems like this can happen if users provide nonsense for the na_object?!

It would have to be a nonsense string that passes `PyUnicode_Check()` 

You're right that the `-2` return value here is a an API wart. I added it as a way for me to catch programming errors. Maybe I should replace that return case with a debug abort inside the static string library and make it so this only fails if you run out of memory?

Reply:
Yeah, true.  The user would have to try very hard...  In principle tha twould mean we have to hceck the size up-front?
TBH, it is also just a minor wart if we raise a MemoryError rather than a size error in practice.

So OK either way, I though passing `nogil` and setting the error inside might be a way, but not sure it is much better.

Reply:
I'd like to fix this in another PR but will get to it.